Libs/Optimize/ParticleSystem/Constraints.cpp

```diff
@@ -0,0 +1,358 @@
+#include "Constraints.h"
+
+namespace itk
+{
+
+void Constraints::addPlane(const vnl_vector<double> &a, const vnl_vector<double> &b,const vnl_vector<double> &c){
+    // See http://mathworld.wolfram.com/Plane.html, for example
+    vnl_vector<double> q;
+    if (DIMENSION == 3)  q = vnl_cross_3d((b-a),(c-a));
+    else if (DIMENSION == 2)  q = vnl_cross_2d((b-a),(c-a));
+
+    if (q.magnitude() > 0.0)
+    {
+      Eigen::Vector3d qp;
+      q = q/q.magnitude();
+      qp(0) = q[0]; qp(1) = q[1]; qp(2) = q[2];
+      PlaneConstraint plane_c;
+      plane_c.SetPlaneNormal(qp);
+      plane_c.SetPlanePoint(a);
+      planeConsts->push_back(plane_c);
+    }
+  }
+
+void Constraints::addSphere(const vnl_vector_fixed<double, DIMENSION> &v, double r){
+    Eigen::Vector3d c; c(0) = v[0]; c(1) = v[1]; c(2) = v[2];
+    SphereConstraint sphere_c;
+    sphere_c.SetCenter(c);
+    sphere_c.SetRadius(r);
+    sphereConsts->push_back(sphere_c);
+}
+
+bool Constraints::transformConstraints(const vnl_matrix_fixed<double, 4, 4> &Trans){
+    return transformPlanes(Trans) & true;
+}
+
+bool Constraints::transformPlanes(const vnl_matrix_fixed<double, 4, 4> &Trans){
+
+  std::cout << "Transforming planes" << Trans <<  std::endl;
+
+  // Convert vnl_matrix to Eigen Matrix
+  Eigen::Matrix4d trans;
```

Comment:
You can tell Eigen to use map vnl's memory(thus avoiding creating another matrix). [Example here](https://github.com/SCIInstitute/ShapeWorks/blob/ab590c3c39e8b11085c4b437f8cdf4c2497b2644/Libs/Utils/Utils.cpp#L446) 
You will have to reason about whether it is safe to do so(ie: the final returned matrix, if any, should not use memory it doesn't own)

 not a big deal for a 4x4 matrix if its not in a hot loop, but if you do have any large matrix/frequently executed code of this kind I'd recommend changing it